File: backend/services/groupsService.go
package services

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"log"
	"run-goals/daos"
	"run-goals/dto"
	"run-goals/models"
	"time"
)

type GroupServiceInterface interface {
	CreateGroup(request dto.CreateGroupRequest) (int64, error)
	UpdateGroup(group dto.UpdateGroupRequest) error
	DeleteGroup(groupID int64) error

	CreateGroupMember(member dto.CreateGroupMemberRequest) error
	UpdateGroupMember(member dto.UpdateGroupMemberRequest) error
	DeleteGroupMember(userID int64, groupID int64) error
	GetGroupMembers(groupID int64) ([]models.GroupMember, error)
	GetGroupMembersGoalContribution(groupID int64, startDate time.Time, endDate time.Time) ([]models.GroupMemberGoalContribution, error)

	CreateGroupGoal(goal dto.CreateGroupGoalRequest) (int64, error)
	UpdateGroupGoal(goal dto.UpdateGroupGoalRequest) error
	DeleteGroupGoal(goalID int64) error

	GetUserGroups(userID int64) ([]models.Group, error)
	GetGroupGoals(groupID int64) ([]models.GroupGoal, error)
	GetGroupGoalByID(goalID int64) (*models.GroupGoal, error)
}

type GroupsService struct {
	l         *log.Logger
	groupsDao *daos.GroupsDao
}

func NewGroupsService(
	l *log.Logger,
	groupsDao *daos.GroupsDao,
) *GroupsService {
	return &GroupsService{
		l:         l,
		groupsDao: groupsDao,
	}
}

func (s *GroupsService) CreateGroup(name string, userID int64) (*int64, error) {
	code, err := s.GenerateGroupCode()
	if err != nil {
		return nil, err
	}
	group := models.Group{
		ID:        0,
		Name:      name,
		Code:      code,
		CreatedBy: userID,
		CreatedAt: time.Now(),
	}
	groupID, err := s.groupsDao.CreateGroup(group)
	if err != nil {
		s.l.Printf("Error calling groupsDao.CreateGroup: %v", err)
		return nil, err
	}

	groupMember := models.GroupMember{
		GroupID:  *groupID,
		UserID:   group.CreatedBy,
		Role:     "admin",
		JoinedAt: group.CreatedAt,
	}
	err = s.groupsDao.CreateGroupMember(groupMember)
	if err != nil {
		s.l.Printf("Error calling groupsDao.CreateGroupMember: %v", err)
		return nil, err
	}

	return groupID, nil
}

func (s *GroupsService) UpdateGroup(groupID int64, name string) error {
	err := s.groupsDao.UpdateGroup(groupID, name)
	if err != nil {
		s.l.Printf("Error calling groupsDao.UpdateGroup: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) DeleteGroup(groupID int64) error {
	err := s.groupsDao.DeleteGroup(groupID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.DeleteGroup: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) CreateGroupMember(groupCode string, userID int64, role string) error {
	id, err := s.groupsDao.GetGroupIDFromCode(groupCode)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetGroupIDFromCode: %v", err)
		return err
	}
	groupMember := models.GroupMember{
		ID:       0,
		GroupID:  *id,
		UserID:   userID,
		Role:     role,
		JoinedAt: time.Now(),
	}
	err = s.groupsDao.CreateGroupMember(groupMember)
	if err != nil {
		s.l.Printf("Error calling groupsDao.CreateMember: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) UpdateGroupMember(groupID int64, userID int64, role string) error {
	err := s.groupsDao.UpdateGroupMember(groupID, userID, role)
	if err != nil {
		s.l.Printf("Error calling groupsDao.UpdateGroupMember: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) DeleteGroupMember(userID int64, groupID int64) error {
	err := s.groupsDao.DeleteGroupMember(userID, groupID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.DeleteGroupMember: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) CreateGroupGoal(request dto.CreateGroupGoalRequest) (*int64, error) {
	// Validate goal type
	validGoalTypes := map[string]bool{
		"distance":         true,
		"elevation":        true,
		"summit_count":     true,
		"specific_summits": true,
	}

	if !validGoalTypes[request.GoalType] {
		return nil, fmt.Errorf("invalid goal type: %s", request.GoalType)
	}

	// Validate specific summits goal
	if request.GoalType == "specific_summits" {
		if len(request.TargetSummits) == 0 {
			return nil, fmt.Errorf("target_summits must be provided for specific_summits goal type")
		}
		// Set target_value to the number of summits for consistency
		request.TargetValue = float64(len(request.TargetSummits))
	}

	goal := models.GroupGoal{
		ID:            0,
		GroupID:       request.GroupID,
		Name:          request.Name,
		Description:   request.Description,
		GoalType:      request.GoalType,
		TargetValue:   request.TargetValue,
		TargetSummits: request.TargetSummits,
		StartDate:     request.StartDate,
		EndDate:       request.EndDate,
		CreatedAt:     time.Now(),
	}

	goalID, err := s.groupsDao.CreateGroupGoal(goal)
	if err != nil {
		s.l.Printf("Error calling groupsDao.CreateGroupGoal: %v", err)
		return nil, err
	}
	return goalID, nil
}

func (s *GroupsService) UpdateGroupGoal(request dto.UpdateGroupGoalRequest) error {
	// Validate goal type
	validGoalTypes := map[string]bool{
		"distance":         true,
		"elevation":        true,
		"summit_count":     true,
		"specific_summits": true,
	}

	if !validGoalTypes[request.GoalType] {
		return fmt.Errorf("invalid goal type: %s", request.GoalType)
	}

	// Validate specific summits goal
	if request.GoalType == "specific_summits" {
		if len(request.TargetSummits) == 0 {
			return fmt.Errorf("target_summits must be provided for specific_summits goal type")
		}
		// Set target_value to the number of summits for consistency
		request.TargetValue = float64(len(request.TargetSummits))
	}

	goal := models.GroupGoal{
		ID:            request.ID,
		GroupID:       request.GroupID,
		Name:          request.Name,
		Description:   request.Description,
		GoalType:      request.GoalType,
		TargetValue:   request.TargetValue,
		TargetSummits: request.TargetSummits,
		StartDate:     request.StartDate,
		EndDate:       request.EndDate,
	}

	err := s.groupsDao.UpdateGroupGoal(goal)
	if err != nil {
		s.l.Printf("Error calling groupsDao.UpdateGroupGoal: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) DeleteGroupGoal(goalID int64) error {
	err := s.groupsDao.DeleteGroupGoal(goalID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.DeleteGroupGoal: %v", err)
		return err
	}
	return nil
}

func (s *GroupsService) GetUserGroups(userID int64) ([]models.Group, error) {
	userGroups, err := s.groupsDao.GetUserGroups(userID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetUserGroups: %v", err)
		return nil, err
	}
	return userGroups, nil
}

func (s *GroupsService) GetGroupGoals(groupID int64) ([]models.GroupGoal, error) {
	groupGoals, err := s.groupsDao.GetGroupGoals(groupID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetGroupGoals: %v", err)
		return nil, err
	}
	return groupGoals, nil
}

func (s *GroupsService) GetGroupMembersGoalContribution(groupID int64, startDate time.Time, endDate time.Time) ([]models.GroupMemberGoalContribution, error) {
	groupMembersContribution, err := s.groupsDao.GetGroupMembersGoalContribution(groupID, startDate, endDate)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetGroupMembersGoalContribution: %v", err)
		return nil, err
	}
	return groupMembersContribution, nil
}

func (s *GroupsService) GetGroupMembers(groupID int64) ([]models.GroupMember, error) {
	groupMembers, err := s.groupsDao.GetGroupMembers(groupID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetGroupMembers: %v", err)
		return nil, err
	}
	return groupMembers, nil
}

func (s *GroupsService) GenerateGroupCode() (string, error) {
	maxAttempts := 5
	for i := 0; i < maxAttempts; i++ {
		bytes := make([]byte, 3)
		_, err := rand.Read(bytes)
		if err != nil {
			return "", err
		}
		code := hex.EncodeToString(bytes)
		exists, err := s.CheckGroupCodeExists(code)
		if err != nil {
			return "", err
		}
		if !exists {
			return code, nil
		}
	}
	return "", fmt.Errorf("Failed to generate unique group code after reaching max attemps: %v", maxAttempts)
}

func (s *GroupsService) CheckGroupCodeExists(code string) (bool, error) {
	count, err := s.groupsDao.CheckGroupCodeExists(code)
	if err != nil {
		s.l.Printf("Error calling groupsDao.CheckGroupCodeExists: %v", err)
		return false, err
	}
	return *count > 0, nil
}

func (s *GroupsService) GetGroupGoalByID(goalID int64) (*models.GroupGoal, error) {
	groupGoal, err := s.groupsDao.GetGroupGoalByID(goalID)
	if err != nil {
		s.l.Printf("Error calling groupsDao.GetGroupGoalByID: %v", err)
		return nil, err
	}
	return groupGoal, nil
}

File: frontend/strava-goal/src/app/services/groups.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, signal } from '@angular/core';
import { Observable } from 'rxjs';
import { DatePipe } from '@angular/common';

@Injectable({
  providedIn: 'root',
})
export class GroupService {
  groups = signal<Group[]>([]);
  groupUpdate = signal<number | null>(null);
  selectedGroup = signal<Group | null>(null);
  goals = signal<Goal[]>([]);
  goalCreated = signal<number | null>(null);
  selectedGoal = signal<Goal | null>(null);
  selectedGoalChange = signal<boolean>(false);
  membersContribution = signal<MemberContribution[]>([]);
  memberAddedOrRemoved = signal<boolean>(false);

  constructor(private http: HttpClient, private datePipe: DatePipe) {}

  createGroup(request: CreateGroupRequest): Observable<CreateGroupResponse> {
    console.log('create group request: ', request);
    return this.http.post<CreateGroupResponse>('/api/groups', request);
  }

  updateGroup(request: UpdateGroupRequest): Observable<any> {
    return this.http.put<any>('/api/groups', request);
  }

  getGroups(): Observable<GetGroupsResponse> {
    return this.http.get<GetGroupsResponse>('/api/groups');
  }

  createGoal(request: CreateGoalRequest): Observable<CreateGoalResponse> {
    return this.http.post<CreateGoalResponse>('/api/group-goal', request);
  }

  updateGoal(request: UpdateGoalRequest): Observable<any> {
    return this.http.put<any>('/api/group-goal', request);
  }

  getGroupGoals(groupID: number): Observable<GetGroupGoalsResponse> {
    const params = new HttpParams().set('groupID', groupID);
    return this.http.get<GetGroupGoalsResponse>('/api/group-goals', { params });
  }

  getGroupMembers(groupID: number): Observable<GetGroupMembersResponse> {
    const params = new HttpParams().set('groupID', groupID);
    return this.http.get<GetGroupMembersResponse>('/api/group-members', {
      params,
    });
  }

  createGroupMember(request: CreateGroupMemberRequest): Observable<any> {
    return this.http.post<any>('/api/group-member', request);
  }

  leaveGroup(request: LeaveGroupRequest): Observable<any> {
    const params = new HttpParams().set('groupID', request.groupID);
    return this.http.delete<any>('/api/group-member', { params });
  }

  getGroupMembersGoalContribution(
    groupID: number,
    startDate: string,
    endDate: string
  ): Observable<GetGroupMembersGoalContributionResponse> {
    const formattedStartDate = this.datePipe.transform(startDate, 'yyyy-MM-dd');
    const formattedEndDate = this.datePipe.transform(endDate, 'yyyy-MM-dd');
    if (formattedStartDate && formattedEndDate) {
      const params = new HttpParams()
        .set('groupID', groupID)
        .set('startDate', formattedStartDate)
        .set('endDate', formattedEndDate);
      return this.http.get<GetGroupMembersGoalContributionResponse>(
        '/api/group-members-contribution',
        { params }
      );
    } else {
      console.log('Failed to format date');
      throw new Error('Failed to format date');
    }
  }

  loadGroups() {
    this.getGroups().subscribe({
      next: (response) => {
        this.groups.set(response.groups);
      },
      error: (err) => {
        console.error('Failed to load groups', err);
      },
    });
  }

  loadGoals(groupID: number) {
    this.getGroupGoals(groupID).subscribe({
      next: (response) => {
        // Don't convert dates to Date objects - keep them as strings for consistency
        const goals = response.goals.map((goal) => ({
          ...goal,
          // Keep dates as strings since the rest of the app expects strings
          start_date: goal.start_date,
          end_date: goal.end_date,
          created_at: goal.created_at,
        }));

        this.goals.set(goals);

        // Find and maintain the currently selected goal after update
        const currentSelectedGoal = this.selectedGoal();
        if (currentSelectedGoal) {
          const updatedSelectedGoal = goals.find(
            (goal) => goal.id === currentSelectedGoal.id
          );
          if (updatedSelectedGoal) {
            this.selectedGoal.set(updatedSelectedGoal);
          }
        }

        // Handle goal creation selection
        const createdGoalId = this.goalCreated();
        if (createdGoalId) {
          const createdGoal = goals.find((goal) => goal.id === createdGoalId);
          if (createdGoal) {
            this.selectedGoal.set(createdGoal);
            this.resetGoalCreated();
          }
        } else if (response.goals.length > 0 && !this.selectedGoal()) {
          // Only set first goal as selected if no goal is currently selected
          this.selectedGoal.set(goals[0]);
        }

        this.notifySelectedGoalChange();
      },
      error: (err) => {
        console.error('Failed to load group goals', err);
      },
    });
  }

  loadMembersGoalContribution(
    groupID: number,
    startDate: string,
    endDate: string
  ) {
    this.getGroupMembersGoalContribution(groupID, startDate, endDate).subscribe(
      {
        next: (response) => {
          this.membersContribution.set(response.members);
        },
        error: (err) => {
          console.error('Failed to load members goal contribution', err);
        },
      }
    );
  }

  notifyGoalCreated(goalID: number) {
    this.goalCreated.set(goalID);
  }
  resetGoalCreated() {
    this.goalCreated.set(null);
  }

  notifySelectedGoalChange() {
    this.selectedGoalChange.set(true);
  }
  resetSelectedGoalChange() {
    this.selectedGoalChange.set(false);
  }

  notifyGroupUpdate() {
    this.loadGroups();
  }
  resetGroupUpdate() {
    this.groupUpdate.set(null);
  }

  notifyMemberAddedOrRemoved(groupMemberID: number) {
    this.memberAddedOrRemoved.set(true);
  }
  resetMemberAddedOrRemoved() {
    this.memberAddedOrRemoved.set(false);
  }

  setSelectedGoal(goal: Goal | null): void {
    this.selectedGoal.set(goal);
    this.notifySelectedGoalChange();
  }

  refreshGoals(groupID: number) {
    this.loadGoals(groupID);
  }

  notifyGoalUpdated(goalID: number) {
    const selectedGroup = this.selectedGroup();
    if (selectedGroup) {
      this.loadGoals(selectedGroup.id);
    }
  }

  deleteGoal(goalId: number): Observable<any> {
    const params = new HttpParams().set('goalID', goalId);
    return this.http.delete<any>('/api/group-goal', { params });
  }

  notifyGoalDeleted(goalId: number) {
    const selectedGroup = this.selectedGroup();
    if (selectedGroup) {
      // If the deleted goal was selected, clear the selection
      if (this.selectedGoal()?.id === goalId) {
        this.selectedGoal.set(null);
      }
      this.loadGoals(selectedGroup.id);
    }
  }
}

export interface Group {
  id: number;
  name: string;
  code: string;
  created_by: number;
  created_at: string;
}

export interface Goal {
  id: number;
  group_id: number;
  name: string;
  description?: string;
  goal_type: 'distance' | 'elevation' | 'summit_count' | 'specific_summits';
  target_value: number; // For distance/elevation/summit count
  target_summits?: number[]; // Array of peak IDs for specific summits goal
  start_date: string;
  end_date: string;
  created_at: string;
  updated_at?: string;
}

export interface Member {
  id: number;
  group_id: number;
  user_id: number;
  role: string;
  joined_at: string;
}

export interface MemberContribution {
  group_member_id: number;
  group_id: number;
  user_id: number;
  role: string;
  joined_at: string;
  total_activities: number | null;
  total_distance: number | null;
  total_unique_summits: number | null;
  total_summits: number | null;
}

export interface CreateGroupRequest {
  name: string;
}

export interface CreateGroupResponse {
  group_id: number;
}

export interface UpdateGroupRequest {
  id: number;
  name: string;
}

export interface GetGroupsResponse {
  groups: Group[];
}

export interface CreateGoalRequest {
  group_id: number;
  name: string;
  description?: string;
  goal_type: 'distance' | 'elevation' | 'summit_count' | 'specific_summits';
  target_value: number;
  target_summits?: number[];
  start_date: string;
  end_date: string;
}

export interface CreateGoalResponse {
  goal_id: number;
}

export interface UpdateGoalRequest {
  id: number;
  group_id: number;
  name?: string;
  description?: string;
  goal_type?: 'distance' | 'elevation' | 'summit_count' | 'specific_summits';
  target_value?: number;
  target_summits?: number[];
  start_date?: string;
  end_date?: string;
}

export interface GetGroupGoalsResponse {
  goals: Goal[];
}

export interface GetGroupMembersGoalContributionResponse {
  members: MemberContribution[];
}

export interface GetGroupMembersResponse {
  members: Member[];
}

export interface CreateGroupMemberRequest {
  group_code: string;
  role: string;
}

export interface LeaveGroupRequest {
  groupID: number;
}
